### PR TITLE
Improve syntax diffing of similar nodes

### PR DESCRIPTION
**Pull request by svick**
Contributes to https://github.com/dotnet/roslyn/issues/17498.

When `SyntaxDiffer` considers whether to look at two non-identical nodes more carefully by "reducing" them, it only considers nodes that have the same `RawKind`. But this means the diff does not work well when a node changes its kind, but otherwise stays mostly the same. This can happen for example when a `class` is changed to a `struct` or when a `Sub` is changed to a `Function`.

This PR improves that by considering each node kind within the same "group" (mathematically, they could be considered to be equivalence classes) to be similar. As a result of this, the code will investigate such pairs more closely, by computing their similarity and then possibly by reducing them together.

It's not completely clear to me which node kinds should form these groups. For example, a constructor could be changed to a factory method with mostly the same code. Or a property could be changed to a method. Does this mean they all should be considered similar?

What I did for now is to only consider the cases where I think there is a reasonably high chance of kind change with few other changes:

* `class`, `struct` and `interface` in C#; `Class`, `Structure`, `Interface` and `Module` in VB
* `Function` and `Sub` in VB

Specifically, for the newly added `TestSyntaxTree_GetChangesWhenSubChangedToFunction`, previously the reported changes were:

```
[12, 57) → "Async Function M() of Task\r\n    Console.WriteLine(42)\r\n  End Function"
```

With this PR, they are instead:

```
[12, 19) → "Async Function M() of Task"
[54, 57) → "Function"
```

As for performance, this PR does mean that similarity will be computed for more pairs of nodes. But since similarity is already computed for example for all pairs of functions (if they are within `MaxSearchLength`), I think this should not make it significantly worse.

cc: @CyrusNajmabadi, @tmat

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
> It's not completely clear to me which node kinds should form these groups. For example, a constructor could be changed to a factory method with mostly the same code. Or a property could be changed to a method. Does this mean they all should be considered similar?

A good starting point would be nodes that have the same *Node class* even if they have different kinds.  So, Classes/Structs/Interfaces have different kinds, but are all TypeDeclarationSyntax in C#.

Where is gets a bit fuzzy are the abstract base types like "MemberDeclarations".  Lots of things are member declarations.  But you wouldn't really expect a 'field' to become a 'destructor' even though they're both members.

File: src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxNode.cs [hunk 1]
```diff
@@ -460,6 +460,26 @@ public new SyntaxTrivia FindTrivia(int position, bool findInsideTrivia = false)
...
+                    case SyntaxKind.InterfaceDeclaration:
+                        return 1;
+                    default:
+                        return -1;
+                }
+            }
+
+            int thisGroup = kindGroup(this);
+
+            return thisGroup != -1 && thisGroup == kindGroup(other);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxNode.cs]: consider checking if `other.GetType() == this.GetType()` instead.

**Comment by svick** [comment]
@CyrusNajmabadi But [`TypeDeclarationSyntax` is also an `abstract` base class](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.syntax.typedeclarationsyntax?view=roslyn-dotnet) for `ClassDeclarationSyntax`, `InterfaceDeclarationSyntax` and `StructDeclarationSyntax`. So using `GetType()` wouldn't work there.

Do you think *adding* a `GetType()` check would be worth it for the cases that actually behave the way you described (e.g. [`BinaryExpressionSyntax`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.syntax.binaryexpressionsyntax?view=roslyn-dotnet))?

**Comment by svick** [comment]
@CyrusNajmabadi Done.

File: src/Compilers/VisualBasic/Test/Syntax/TestSyntaxNodes.vb [hunk 2]
```diff
@@ -2796,6 +2796,25 @@ End Module
             Assert.Throws(Of ArgumentNullException)(Sub() newTree.GetChangedSpans(BlankTree))
         End Sub
 
+        <Fact>
+        Public Sub TestSyntaxTree_GetChangesWhenSubChangedToFunction()
+            Dim parseLines = 
+                Function(lines As String()) SyntaxFactory.ParseSyntaxTree(String.Join(vbCrLf, lines))
+
+            Dim oldTree = parseLines({ "Module M", "  Sub M()", "    Console.WriteLine(42)", "  End Sub", "End Module"})
+            Dim newTree = parseLines({ "Module M", "  Async Function M() of Task", "    Console.WriteLine(42)", "  End Function", "End Module"})
```
Comment (svick) [inline src/Compilers/VisualBasic/Test/Syntax/TestSyntaxNodes.vb]: When I used verbatim strings for code, this test failed on Unix, because of different line endings, which in turn mean different `TextSpan`s.

So I switched to this approach, which ensures `\r\n` is used on any OS. Is there a better way to do this?

**Comment by jinujoseph** [comment]
@sharwell to review this first from IDE team

File: src/Compilers/Core/Portable/Syntax/SyntaxNode.cs [hunk 3]
```diff
@@ -1221,6 +1221,29 @@ internal SyntaxNode GetNodeWithoutAnnotations(IEnumerable<SyntaxAnnotation> anno
...
+                return true;
+            }
+
+            if (this.IsSimilarToCore(other))
+            {
+                return true;
+            }
+
+            return false;
+
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/SyntaxNode.cs]: extra newline.

File: src/Compilers/VisualBasic/Portable/Syntax/VisualBasicSyntaxNode.vb [hunk 4]
```diff
@@ -474,6 +474,24 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             Return SyntaxTree.GetDiagnostics(Me)
         End Function
 
+        Private Protected Overrides Function IsSimilarToCore(other As SyntaxNode) As Boolean
```
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Portable/Syntax/VisualBasicSyntaxNode.vb]: i'm a bit wary of these going directly on SyntaxNode.  i feel like it would be better to just be in the syntaxdiffer.  that may be a bit annoying as we'd need to make SyntaxDiffer inheritable, but i do think like it would be better to isolate tehre rather than have this particular one-off here. 

Particularly: the primary issue i have is that SyntaxNode isn't wishy-washy :)  It's a strictly defined, highly invariant-based API.  I'd like it to stay that way.  Well defined, with clear semantics on how every part of it should behave.  However, SyntaxDiffer, but it's very nature is going to make a 'best effort' on doing a diff since htere's no perfect algorithm out there.  So having this sort of logic go in there just seems more sensible to me.

**Comment by tmat** [comment]
I'm not sure this is the right approach. We already have a much more sophisticated syntax differ here http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/EditAndContinue/TopSyntaxComparer.cs
currently used for Edit and Continue. I don't think we should make the other API more complicated. Instead, if smarter differ is needed in some scenario it could perhaps switch to the EnC tree differ.

**Comment by CyrusNajmabadi** [comment]
I think a core problem isthat TopSytnaxComparer is internal (right)?  So when customers call into our public diffing points, they're basically only going to get the current behavior which can sometimes be mediocre.  I'm personally ok with targetted fixes to the current system.  I don't believe it was ever intended to be 'done', it just tries to not be overly complicated so that it has reasonable perf characteristics.

Note: if the EnC differ is both good *and* fast, i would be totally ok with it just becoming the defacto differ (or maybe something that can be opted into if we're worried about breaking behavior).

--

Basically, if @svick was making thousands of lines of changes here, i'd be super wary about things.  But if this is jsut a couple of targeted fixes to improve behavior in a manner customers would fine sensible, then i think the investment is fine.

**Comment by CyrusNajmabadi** [comment]
Also, i would not let hte perfect be the enemy of the good.  I would both support a longer term plan to make it possible to use the EnC differ, while also supporting short term easy fixes that improve real user scenarios today with our existing public diff engine.  No reason we can't have both! :)

**Comment by heejaechang** [comment]
@mavasani I believe you have looked at this code before, can you take a look?

by the way, I believe this differ is first introduced to get text changes for compiler incremental parsing (ex, IDE did tree transformation and want to apply changes to workspace which requires to get text changes out of the tree transformation. and it only required to be good enough for incremental parsing) and code is written for that. and later, since it already existed, IDE starts to use it for all other cases such as fix all, or showing diff to users in preview window and etc.

but since it was written for other purpose (especially for speed), it wasn't exactly suitable for what IDE wanted.

for IDE usage, we used to have an issue saying refactoring differ @tmat created for Enc and use that for IDE usage rather than this one.

the advantage of the Enc differ is that, one can set tree level where the diff happens. 

for example, for rename or signature change
image

when show preview reference changes is set, it was a common bug where we showed text level edit as changes rather than syntactical level such as telling users whole identifier is changed, or method decl is changed, instead we show as a word being changed or a line as changed.

I am +1 on IDE moving to Enc differ eventually so that we can have much flexible way to diff trees based on usage cases.

File: src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxNode.cs [hunk 5]
```diff
@@ -460,6 +460,26 @@ public new SyntaxTrivia FindTrivia(int position, bool findInsideTrivia = false)
 
 #region SyntaxNode members
 
+        private protected override bool IsSimilarToCore(SyntaxNode other)
+        {
+            int kindGroup(SyntaxNode node)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxNode.cs]: >kindGroup [](start = 16, length = 9)

nit: we tend to put local functions last. I went ahead and adjusted to save time

File: src/Compilers/VisualBasic/Test/Syntax/TestSyntaxNodes.vb (hunk 2 above)
Comment (jcouv) [reply src/Compilers/VisualBasic/Test/Syntax/TestSyntaxNodes.vb]: Did the pattern used by other `GetChanges` tests (such as `TestSyntaxTree_GetChangesInValid`) cause the same problem?

---
In reply to: [226878038](https://github.com/dotnet/roslyn/pull/30634#discussion_r226878038) [](ancestors = 226878038)

**Comment by jaredpar** [comment]
This PR appears to have gotten stale (last OP interaction was mid October). Going to close this out for now. Can re-open if the OP wants to proceed here.

File: src/Compilers/VisualBasic/Portable/Syntax/VisualBasicSyntaxNode.vb (hunk 4 above)
Comment (svick) [reply src/Compilers/VisualBasic/Portable/Syntax/VisualBasicSyntaxNode.vb]: @CyrusNajmabadi I would like to come back to this PR, but I'm having trouble figuring out a good way to separate this code from the language-specific `SyntaxNode`, while keeping it in the language-specific assembly (because it needs access to `SyntaxKind`).

The best approach I could think of is if `SyntaxNode` gave access to the correct `SyntaxDiffer`. In other words, `CSharpSyntaxNode` would contain:

```c#
internal override SyntaxDiffer GetDiffer(SyntaxNode newNode, bool computeNewText) =>
    new CSharpSyntaxDiffer(this, newNode, computeNewText);
```

Do you think this would be a good solution? If not, what do you suggest?

Comment (CyrusNajmabadi) [reply src/Compilers/VisualBasic/Portable/Syntax/VisualBasicSyntaxNode.vb]: @svick That seems reasonable to me.  You can have an abstract base class with core logic, and any lang specific logic can go in the derived types.

**Comment by svick** [comment]
I apologize I have been silent for so long.

I would like to proceed with this PR and for that purpose I have now rebased the branch on current master and added a commit which should resolve https://github.com/dotnet/roslyn/pull/30634#discussion_r237609197 (though the changes don't seem to show up here right now). Can this PR be reopened and re-reviewed? Thanks.

File: src/Compilers/VisualBasic/Test/Syntax/TestSyntaxNodes.vb (hunk 2 above)
Comment (svick) [reply src/Compilers/VisualBasic/Test/Syntax/TestSyntaxNodes.vb]: No, because they're not comparing a multi-line string against a constant `TextSpan`. So this seems to be the only test which has this problem.

**Comment by RikkiGibson** [comment]
The reopen button is disabled for me.

Screen Shot 2019-08-15 at 9 53 26 AM

It's possible you may need to open a new PR and link to this one from it. Sorry about the inconvenience.

BTW, we typically ask that people avoid rebasing/force pushing PRs that are not drafts or that have already been reviewed.

**Comment by svick** [comment]
@RikkiGibson 

> It's possible you may need to open a new PR and link to this one from it. Sorry about the inconvenience.

I didn't know it worked like that, thanks for letting me know. I have opened it as https://github.com/dotnet/roslyn/pull/38016.

> we typically ask that people avoid rebasing/force pushing PRs that are not drafts or that have already been reviewed.

I thought that since it was such a long time, rebasing will be necessary, because there would be some conflicts. But I actually didn't have to resolve any conflicts manually when I did the rebase, so yeah, I shouldn't have done that rebase before knowing it was necessary.
